### PR TITLE
Fix parameter -Algorithm & -ExcludeAlgorithm

### PR DESCRIPTION
**Pull request by UselessGuru**
### Please review and approve. This is a prerequisite for ExcavatorNHMP.

The 'artificial' algo names like ...-NHMP do not work with the current implementation

This is a prerequisite for ExcavatorNHMP.

### CODE REVIEW

**Comment by UselessGuru** [comment]
@aaronsace 
 > Please allow both full and partial matches

How should that work?
Assuming I have
 -algorithm Equihash
should it allow Equihash AND Equihash-NHMP or only Equihash?

I cannot see how to implement this. Sorry.

**Comment by aaronsace** [comment]
@UselessGuru Good spot, `Get-Algorithm` will need to be run on the first part then rejoined. It should then work how sub-names were meant to; it's a reason why hyphens and underscores were reserved. 

P.S. maybe add `-Unique` to `Select-Object`

**Comment by UselessGuru** [comment]
> will need to be run on the first part then rejoined.

???

File: MultiPoolMiner.ps1 [hunk 1]
```diff
@@ -343,8 +343,7 @@ while ($true) {
             ForEach-Object {if (-not $_.DeviceName) {$_ | Add-Member DeviceName (Get-Device $_.Type).Name -Force}; $_} | #for backward compatibility
             Where-Object {$_.DeviceName} | #filter miners for non-present hardware
             Where-Object {(Compare-Object @($Devices.Name | Select-Object) @($_.DeviceName | Select-Object) | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0} | 
-            Where-Object {($Config.Algorithm.Count -eq 0 -or (Compare-Object $Config.Algorithm $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0) -and ((Compare-Object $Pools.PSObject.Properties.Name $_.HashRates.PSObject.Properties.Name | Where-Object SideIndicator -EQ "=>" | Measure-Object).Count -eq 0)} |
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: $Config.Algorithm & $Config.ExcludeAlgorithm camparison is not needed here.
Lines 297 & 298 allow only matching pools.
New line 346 allows only miners for which we have pools.

File: MultiPoolMiner.ps1 [hunk 2]
```diff
@@ -451,7 +450,7 @@ while ($true) {
...
+    if ($Config.UseFastestMinerPerAlgoOnly) {$Miners = $Miners | Sort-Object -Descending {"$($_.DeviceName -join '')$(($_.HashRates.PSObject.Properties.Name | Foreach-Object {$_ -split "-" | Select-Object -Index 0}) -join '')$(if($_.HashRates.PSObject.Properties.Value -eq $null) {$_.Name})"}, {($_ | Where-Object Profit -EQ $null | Measure-Object).Count}, {([Double]($_ | Measure-Object Profit_Bias -Sum).Sum)}, {($_ | Where-Object Profit -NE 0 | Measure-Object).Count} | Group-Object {"$($_.DeviceName -join '')$(($_.HashRates.PSObject.Properties.Name | ForEach-Object {$_ -split "-" | Select-Object -Index 0}) -join '')$(if($_.HashRates.PSObject.Properties.Value -eq $null) {$_.Name})"} | Foreach-Object {$_.Group[0]}}
```
Comment (UselessGuru) [inline MultiPoolMiner.ps1]: [AlgoName] & [AlgoName]-NHMP are treated as same algo

**Comment by aaronsace** [comment]
@UselessGuru I meant this line: 
https://github.com/MultiPoolMiner/MultiPoolMiner/blob/aa6289ef4875c97ccbe80426a9f1f862a1f380d4/MultiPoolMiner.ps1#L87

-split; select -index 0; Get-Algo; -join

**Comment by UselessGuru** [comment]
> -split '-'; select -index 0; Get-Algo; -join '-'

There is no need for this because we've removed all hyphens and underscores in algorithms.txt

What else needs to be changed until you approve this PR?

**Comment by UselessGuru** [comment]
@aaronsace 
Could you please clarify this - thank you! We should get this approved.

**Comment by UselessGuru** [comment]
@aaronsace 

I think you need to approve this. I'ts looking good for me.

**Comment by mkirov0** [comment]
Sorry if this sounds like a dumb question but will this allow me to use "ExcludeAlgorithm" on per pool base in the config file?
Like:
"BlockMasters": {
      "BTC": "$Wallet",
      "Worker": "$WorkerName",
      "ExcludeAlgorithm": ["algo1", "alg2", "algo3"]

**Comment by UselessGuru** [comment]
@mkirov0 

> will this allow me to use "ExcludeAlgorithm" on per pool base 

No. it will not yet. But you raise a good point. I'll create a new PR for this feature after this PR is approved.

**Comment by mkirov0** [comment]
@UselessGuru I added some lines from ZergPool.ps1 to Zpool.ps1 and BlockMasters.ps1. It started polling the individual coins for me just like ZergPool.ps1 does. I also tried to work on AHashPool.ps1, but I think their API standard reports in a slightly different fashion... I did not want to create PR, but I am uploading the files here if you want to take a look and possibly use in the future. Thanks for all the work you guys do!
[Archive.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2200596/Archive.zip)

**Comment by UselessGuru** [comment]
@mkirov0

I am working on more standardized pool files. The attached pool files are a full replacement for the existing ones. If thei work I might create a PR for them
[Pools.zip](https://github.com/MultiPoolMiner/MultiPoolMiner/files/2200848/Pools.zip)

Note: BlazePool does not expose the coin info via API. And not all pools expose the same properties, e.g. workers, through their API.

File: MultiPoolMiner.ps1 (hunk 2 above)
Comment (aaronsace) [reply MultiPoolMiner.ps1]: Why is this whole line even here? What's wrong with `$BestMiners`?

File: MultiPoolMiner.ps1 (hunk 1 above)
Comment (aaronsace) [reply MultiPoolMiner.ps1]: I would rather leave this in as you know from this point that all the filters have been applied.

File: MultiPoolMiner.ps1 (hunk 2 above)
Comment (UselessGuru) [reply MultiPoolMiner.ps1]: $BetMiners is just THE BEST miner per device and lists only ONE algo.

The whole line creates a list of the all fastest miners listing all available algos.

**Comment by UselessGuru** [comment]
Fixed with https://github.com/MultiPoolMiner/MultiPoolMiner/pull/2094

**Pull request closed without merging**